CORE-8619 crash_tracker: print recorded crash reports

pgellert opened the pull request:
The PR implements reading recorded crash reports and extending the crash loop limit reached log message with a description of the recorded crashes.

To limit the length of the output, up to the 5 earliest and 5 latest recorded crashes are printed only. A unit test is added to test this logic and the output format.

A ducktape test is added to verify that information about recorded startup exceptions is correctly printed to the logs when the crash loop limit is reached. Through this we test that both:
 * Startup crash information is correctly recorded
 * The recorded crash information is correctly printed

Fixes: https://redpandadata.atlassian.net/browse/CORE-8619

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
crash_tracker: add config as cmake dependency

The limiter uses node configs, so this should be included in the list of
dependencies.

This was missed from cmake only. It is already present in bazel.

pgellert committed:
crash_tracker: implement describing crashes

Implement the logic of formatting a list of crashes to a string. This is
going to be printed to the logs when the crash loop limit is reached.

See the added unit tests for what the output is going to look like.

pgellert committed:
crash_tracker: gather and print recorded crashes

... when the crash loop limit is reached.

pgellert committed:
dt/crash_tracker: verify crash log message

Verify that information about recorded startup exceptions is correctly
printed to the logs when the crash loop limit is reached. Through this
we test that both:
 * Startup crash information is correctly recorded
 * The recorded crash information is correctly printed

michael-redpanda commented on src/v/crash_tracker/types.cc:
```diff
@@ -13,6 +13,24 @@
+std::ostream& operator<<(std::ostream& os, const crash_description& cd) {
+    fmt::print(os, "{}", cd.crash_message.c_str());
+
+    const auto opt_stacktrace = cd.stacktrace.c_str();
+    const auto has_stacktrace = strlen(opt_stacktrace) > 0;
+    if (has_stacktrace) {
+        fmt::print(os, " Backtrace: {}.", opt_stacktrace);
+    }
```
why not use `cd.stacktrace.size()`?

pgellert replied:
It's because when we don't write anything to `stacktrace` or `addition_info`, their size is 4096 because they contain 4096 '\0' characters.

This was a bit... brittle. So I decided now to wrap these reserved strings into a `reserved_string` class. This documents the distinction between getting the length of the written data vs getting the capacity of the string.

michael-redpanda replied:
> It's because when we don't write anything to stacktrace or addition_info, their size is 4096 because they contain 4096 '\0' characters.

Ah my bad, I forgot that's what `size()` returned, I thought it looked for null characters but it doesn't.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--705ea284-da74-4dd5-9065-5f001bf59d7a-->
<!--TEST RESULTS-->
<!--d4fb6409-d9c6-48e7-91b9-a8597c208a47-->
<!--TEST RESULTS-->
<!--91e624cb-685f-4827-9a9b-e2238529f341-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61040">build#61040</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61040#01948e05-41a8-41be-9919-fb59395d528a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61169">build#61169</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61169#019498ba-4e1c-46fc-8ceb-93e766ac0da7 | FLAKY | 1/2 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=False.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=None.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af8-4623-ae8f-294ec2d1f2f2 | FAIL | 0/20 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=True.use_sasl=True.enable_authz=True.authn_method=mtls_identity.client_auth=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f8-4ff3-852e-62783b044cc5 | FAIL | 0/20 |
| rptest.tests.configuration_update_test.ConfigurationUpdateTest.test_update_invalid_advertised_api | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2af9-4249-bfe1-b775967bd356 | FAIL | 0/20 |
| rptest.tests.configuration_update_test.ConfigurationUpdateTest.test_update_invalid_advertised_api | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45fa-4082-af2c-027c123ced10 | FAIL | 0/20 |
| rptest.tests.strict_data_init_test.StrictDataInitTest.test_strict_data_init_enabled | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#019498ff-2afa-46ac-b1b8-14a553d44d7d | FAIL | 0/20 |
| rptest.tests.strict_data_init_test.StrictDataInitTest.test_strict_data_init_enabled | ducktape | https://buildkite.com/redpanda/redpanda/builds/61169#01949903-45f9-423f-9ba5-c61c43280ef3 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61236">build#61236</a></summary>


| test_id                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61236#0194a97f-d99c-48fd-80f3-8c60e305dad6 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61236#0194a984-55cb-4757-9c9b-b05b7f4efb07 | FLAKY         | 1/2      |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61236#0194a984-55cb-4757-9c9b-b05b7f4efb07 | FLAKY         | 1/2      |
</details>
<!--/body-->

michael-redpanda commented on src/v/crash_tracker/types.cc:
```diff
@@ -13,6 +13,24 @@
+        fmt::print(os, " Backtrace: {}.", opt_stacktrace);
+    }
+
+    const auto opt_add_info = cd.addition_info.c_str();
+    const auto has_add_info = strlen(opt_add_info) > 0;
+    if (has_add_info) {
+        fmt::print(os, " {}", opt_add_info);
+    }
```
same question

pgellert replied:
See https://github.com/redpanda-data/redpanda/pull/24883#discussion_r1925894446

michael-redpanda commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -137,6 +138,37 @@ void recorder::record_crash_exception(std::exception_ptr eptr) {
+        co_return result;
+    }
+
+    for (const auto& entry :
+         std::filesystem::directory_iterator(crash_report_dir)) {
+        if (!entry.path().string().ends_with(crash_report_suffix)) {
+            // Filter only for crash files
+            continue;
+        }
+
+        auto buf = co_await read_fully(entry.path());
+        try {
+            auto crash_desc = serde::from_iobuf<crash_description>(
+              std::move(buf));
+            result.emplace_back(std::move(crash_desc));
+        } catch (const serde::serde_exception&) {
+            vlog(
+              ctlog.warn,
+              "Ignoring malformed crash report file {}",
+              entry.path());
+        }
+    }
```
question: Can this possibly return an unsorted list of reports?  Or does `directory_iterator` guarantee some sort of ordering?

pgellert replied:
Good point. In practice it does look like it's always sorted but the API states that this shouldn't be relied on so I've added a sort here to ensure that these are always returned in increasing time order.

> The iteration order is unspecified, except that each directory entry is visited only once.

https://en.cppreference.com/w/cpp/filesystem/directory_iterator

michael-redpanda reviewed (commented):
couple of questions/nits

pgellert committed:
crash_tracker: introduce `reserved_string`

reserved_string is a simple wrapper around ss::sstring that allows
pre-allocating a string of a certain size with trailing '\0's and allows
safer access to the populated part of the string.

dotnwat commented on src/v/crash_tracker/recorder.h:
```diff
@@ -42,6 +43,14 @@ class recorder {
+    /// ss::util::spinlock instead of more efficient locking mechanisms like the
+    /// seastar primitives or std::mutex to ensure that we can safely use the
+    /// lock in a singal handler context.
+    ss::util::spinlock _writer_lock;
```
can you comment on what you mean by async-signal safe here and what concurrency you are trying to limit, exactly? if the you're holding this lock and the current thread handles a signal that tries to acquire the same lock, then it's going to be a deadlock, no?

maybe we can design this so that we don't need any locking? holding a spinlock while doing I/O and co_awaiting futures doesn't feel like a great idea.

pgellert replied:
Thanks. This part of the change is in the previous PR so let's discuss it there: https://github.com/redpanda-data/redpanda/pull/24854#discussion_r1925869815

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/8e22bb5e038c89d6471066f6b24cc5e5a6c98026..651c5fd990e015c28e16a4ea498c2b8307cc89e5): address a round of review from @michael-redpanda

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/651c5fd990e015c28e16a4ea498c2b8307cc89e5..572ef2dc0cfefbd7aec61cf49b84ef1164535b1a): rebased to `crashlog/record-startup-exc` (no changes to the last 4 commits of this PR)

michael-redpanda commented on src/v/crash_tracker/prepared_writer.cc:
```diff
@@ -0,0 +1,172 @@
+        return nullptr;
+    }
+
+    vassert(_state == state::initialized, "Unexpected state: {}", _state);
```
question: are the states necessary anymore now that we're using atomics?  Or, and maybe this idea is dumb, can you put the enum state in an atomic?  Feels like we're tracking state in two different ways now.

pgellert replied:
Yeah I wasn't happy about the duplication of state tracking here. I changed to an atomic enum now like you suggested in this push to the base PR: https://github.com/redpanda-data/redpanda/pull/24854#issuecomment-2613226164

The (I think acceptable) downside is:
* We rely on `std::atomic<state>` being lock-free, which it should be on the architectures we support now, but perhaps later we might need some changes here if we add support for new architectures.
* We need to pay attention to whether `_state` is updated on entry or exit to the methods of `prepared_writer`, because it matters for the cross-thread visibility requirements. I've added a comment to make note of this.

But overall it does simplify the code so I did move over to the `std::atomic<state>` version.

michael-redpanda commented on src/v/crash_tracker/prepared_writer.cc:
```diff
@@ -0,0 +1,172 @@
+}
+
+ss::future<>
+prepared_writer::initialize(std::filesystem::path crash_file_path) {
+    _crash_report_file_name = std::move(crash_file_path);
+
+    _serde_output.reserve_memory(crash_description::serde_size_overestimate);
+
+    // Create the crash recorder file
+    auto f = co_await ss::open_file_dma(
+      _crash_report_file_name.c_str(),
+      ss::open_flags::create | ss::open_flags::rw | ss::open_flags::truncate
+        | ss::open_flags::exclusive);
+    co_await f.close();
+
+    // Open the crash recorder file using ::open().
+    // We need to use the low level open() function here instead of the seastar
+    // API or higher-level C++ primitives because we need to be able to
+    // manipulate the file using async-signal-safe, allocation-free functions
+    // inside signal handlers.
+    _fd = ::open(_crash_report_file_name.c_str(), O_WRONLY);
+    if (_fd == -1) {
+        throw std::system_error(
+          errno,
+          std::system_category(),
+          fmt::format(
+            "Failed to open {} to record crash reason",
+            _crash_report_file_name));
+    }
+
+    _state = state::initialized;
+    _initialized = true;
+}
```
question: Do you need to check if you've been initialized before doing the initialization?

pgellert replied:
We don't strictly need to, but I'll add it as a sanity check + for code readability.

michael-redpanda commented on src/v/crash_tracker/prepared_writer.cc:
```diff
@@ -0,0 +1,172 @@
+        co_return;
+    }
+
+    vassert(_state == state::initialized, "Unexpected state: {}", _state);
```
is this correct?  Shouldn't it be `filled`?

michael-redpanda replied:
oh n/m I see the updated state machine

michael-redpanda replied:
Actually according to the state machine you can go from `uninitialized` to `released`.  So there's a discrepancy there

pgellert replied:
I've fixed the state machine now. I've pushed the fixes to the base PR here: https://github.com/redpanda-data/redpanda/pull/24854#issuecomment-2613226164

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61169-->
<!--f385ee51-1cf5-41a9-ac1c-f0a4d532d7e7-->
<!--d6477e01-0e0a-4485-b69c-01259db5fcd2-->
<!--8f6c26e1-38db-4cf6-b129-e9748175dcc0-->
<!--7edd4bad-b54c-413b-8604-bb1cb28fb938-->
<!--aa668f99-f65d-4e69-a2dd-09ba74be20be-->
<!--c8e4c12b-8c80-41e2-b0e3-466572748fbf-->
<!--title-->
## Retry command for [Build#61169](https://buildkite.com/redpanda/redpanda/builds/61169)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/strict_data_init_test.py::StrictDataInitTest.test_strict_data_init_enabled
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":false,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":null,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":true,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":false,"use_sasl":true,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":null,"use_sasl":true,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":true,"use_sasl":true,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":false,"use_sasl":false,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":null,"use_sasl":false,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":true,"use_sasl":false,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":false,"use_sasl":true,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":null,"use_sasl":true,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":false,"enable_authz":true,"use_sasl":true,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":false,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":null,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":true,"use_sasl":false,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":false,"use_sasl":true,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":null,"use_sasl":true,"use_tls":false}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"mtls_identity","client_auth":true,"enable_authz":true,"use_sasl":true,"use_tls":false}
tests/rptest/tests/configuration_update_test.py::ConfigurationUpdateTest.test_update_invalid_advertised_api
```
<!--/body-->

michael-redpanda commented:
oh @pgellert I'm sorry I think I commented in the wrong PR

pgellert commented:
No worries. I'll address your comments on https://github.com/redpanda-data/redpanda/pull/24854 and put this one into draft mode until that merges.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/572ef2dc0cfefbd7aec61cf49b84ef1164535b1a..c212703de97199d65170af58a2b11e9f8e6988b7): rebase to dev now that https://github.com/redpanda-data/redpanda/pull/24854 merged
